[SYCL][CUDA][HIP] Device info query for maximum registers per block and targetted exception on out-of-registers for CUDA. (#9106)
intel/llvm: sycl/doc/extensions/experimental/sycl_ext_codeplay_max_registers_per_work_group_query.md

# sycl_ext_codeplay_max_registers_per_work_group_query

## Notice

This document describes an **experimental** API that applications can use to try
out a new feature. Future versions of this API may change in ways that are
incompatible with this experimental version.

## Introduction

> This extension adds functionally a new device information descriptor. It provides the ability to query a device for the maximum number of registers available per work-group.

Reviewer: Nit; I think this reads a little better.
```suggestion
This extension adds a new device information descriptor that provides the ability to query a device for the maximum number of registers available per work-group.
```

Author: Does sound much clearer. Appreciated!

| ------------------------------------------------------ | ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| ext::codeplay::experimental::info::device::max_registers_per_work_group     |  unsigned int      | Returns the maximum number of registers available for use per work-group based on the capability of the device.       |

### Note

## Examples

```c++
sycl::device gpu = sycl::device{sycl::gpu_selector_v};
std::cout << gpu.get_info<sycl::info::device::name>() << '\n';

#ifdef SYCL_EXT_CODEPLAY_MAX_REGISTERS_PER_WORK_GROUP_QUERY
unsigned int registers_per_group = gpu.get_info<sycl::ext::codeplay::experimental::info::device::max_registers_per_work_group>();
std::cout << "Max registers per work-group: " << registers_per_group << '\n';
#endif
```

Ouputs to the console:

Executed using the CUDA back-end on NVIDIA.

```
NVIDIA ...
Max registers per work-group: 65536
```

- See: [CUDA Toolkit Documentation](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capabilities)

> ## Implementation

Reviewer: I would prefer if we just drop this section.

Reviewer: That includes its sub-section(s). 😉

Author: Ah yeah, excuse me. Never good to rush work on Fridays.😅 Cheers for double checking that.
